src/main.rs

```diff
@@ -1,183 +1,153 @@
+                    }
+                },
+                Message::Notification(notification) => {
+                    let _ = self.handle_notification(notification);
+                },
+                Message::Response(_) => (),
             }
         }
     }
-    fn send<T: serde::Serialize + fmt::Debug>(&mut self, msg: &T) -> Result<(), Error> {
-        writeln!(self.log, "Sending: {:?}", msg)?;
-        let bytes = serde_json::to_vec(msg)?;
-        write!(self.stdout, "Content-Type: application/vscode-jsonrpc; charset=utf-8\r\n")?;
-        write!(self.stdout, "Content-Length: {}\r\n", bytes.len())?;
-        write!(self.stdout, "\r\n")?;
-
-        self.stdout.write_all(&bytes)?;
-        self.stdout.flush()?;
-        Ok(())
-    }
     fn handle_request(&mut self, req: Request) -> Result<(), Error> {
         match &*req.method {
-            "initialize" => self.send(&Response::success(req.id, Some(
-                InitializeResult {
-                    capabilities: ServerCapabilities {
-                        text_document_sync: Some(TextDocumentSyncCapability::Options(
-                            TextDocumentSyncOptions {
-                                open_close: Some(true),
-                                change: Some(TextDocumentSyncKind::Full),
-                                ..Default::default()
-                            }
-                        )),
-                        completion_provider: Some(CompletionOptions {
-                            resolve_provider: Some(true),
-                            ..Default::default()
-                        }),
-                        definition_provider: Some(true),
-                        document_formatting_provider: Some(true),
-                        rename_provider: Some(RenameProviderCapability::Simple(true)),
-                        ..Default::default()
-                    }
-                }
-            )))?,
-            "textDocument/didOpen" => {
-                let params: DidOpenTextDocumentParams = serde_json::from_value(req.params)?;
-                let text = params.text_document.text;
-                let parsed = rnix::parse(&text);
-                self.send_diagnostics(params.text_document.uri.clone(), &text, &parsed)?;
-                self.files.insert(params.text_document.uri, (parsed, text));
-            },
-            "textDocument/didChange" => {
-                let params: DidChangeTextDocumentParams = serde_json::from_value(req.params)?;
-                if let Some(change) = params.content_changes.into_iter().last() {
-                    let parsed = rnix::parse(&change.text);
-                    self.send_diagnostics(params.text_document.uri.clone(), &change.text, &parsed)?;
-                    self.files.insert(params.text_document.uri, (parsed, change.text));
-                }
-            },
-            "textDocument/definition" => {
+            GotoDefinition::METHOD => {
```

Comment:
the extraction could be done in a type-safer way: https://github.com/rust-analyzer/rust-analyzer/blob/f90a88630195a1b6b542b5883295d66dd6c59cef/crates/ra_lsp_server/src/main_loop.rs#L521-L540